src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  // Tablica, która będzie przechowywać historię stanów
+  const stateHistory = [];
+
+  // Zaczynamy od klonowania początkowego stanu
+  let currentState = { ...state };
+
+  // Iterujemy po każdej akcji
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        // Tworzymy nowy pusty obiekt stanu
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        // Dodajemy nowe właściwości do sklonowanego stanu
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        // Klonujemy stan i usuwamy wskazane klucze
+        currentState = { ...currentState };
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    // Dodajemy sklonowany i zmodyfikowany stan do historii
+    stateHistory.push({ ...currentState });
```

Comment:
[DONT REPEAT YOURSELF] - Pushing a copy of the object to the `stateHistory` array is correctly done at the end of the loop cycle and outside of the `switch` block, which is good as it avoids repetition and follows the DRY principle.